_posts/2019-04-09-spring-security-in-action.md

```diff
@@ -0,0 +1,224 @@
+    return defaultTokenServices;
+}
+```
+
+4. The authenticationManager() function is defining the token verify logic. This class will be use to check the user authentication when a token is refreshed.
+
+### IV. Optional: custome token authority verification logic
+If the default token manager does not meet your requirement, which is happening all the time, you could assign you own authenticationProvider by following code.
+```java
+@Override
+protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+    auth.authenticationProvider(jwtAuthenticationProvider());
+}
+```
+
+And the provider can be like this
+```java
+@Override
+public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+    DecodedJWT jwt = ((JwtAuthenticationToken) authentication).getToken();
+    //I want a token never expired.
+    //if (jwt.getExpiresAt().before(Calendar.getInstance().getTime()))
+        //throw new NonceExpiredException("Token expires");
+    String clientId = jwt.getSubject();
+    UserDetails user = userService.loadUserByUsername(clientId);
+    if (user == null || user.getPassword() == null)
+        throw new NonceExpiredException("Token expires");
+    Algorithm algorithm = Algorithm.HMAC256(user.getPassword());
+    JWTVerifier verifier = JWT.require(algorithm)
+            .withSubject(clientId)
+            .build();
+    try {
+        verifier.verify(jwt.getToken());
+    } catch (Exception e) {
+        throw new BadCredentialsException("JWT token verify fail", e);
+    }
+    return new JwtAuthenticationToken(user, jwt, user.getAuthorities());
+}
+```
+### V. Optional: custome verify chain
+If the authenticate() function throws any exception, we may to handle it or just record it in the database. To implement it we can just set tokenValidSuccessHandler and tokenValidFailureHandler. In the handler, you can rewrite onAuthenticationSuccess and onAuthenticationFailure with your own logic.
+```java
+...
+    .and()
+        .apply(new MyValidateConfigure<>())
+            .tokenValidSuccessHandler(myVerifySuccessHandler())
+            .tokenValidFailureHandler(myVerifyFailureHandler())
+...
+```
+```java
+public class JwtAuthenticationFailureHandler implements AuthenticationFailureHandler {
+    @Override
+    public void onAuthenticationFailure(HttpServletRequest request, HttpServletResponse response,
+                                        AuthenticationException exception) {
+        response.setStatus(HttpStatus.UNAUTHORIZED.value());
+    }
+}
+```
+
+### VI. UserDetailService
+UserDetailService is the core interface which loads user-specific data. We mast realize loadUserByUsername() function to locate user and user's role.
+
+### VII. Rest APIs
+Once all authoritioin configure has been finished, you can enjoy your developing. To control your role and access you can simply add @PreAuthorize("hasAuthority('ADMIN_USER')") in your rest api declaration.
```

Comment:
Once all `the authorization configurations have` been finished, you can `start to` enjoy `developing your project`. To control ... ....